crates/fuzzy_filter/src/lib.rs

```diff
@@ -83,12 +84,646 @@ pub fn fuzzy_filter_and_rank(
+    number: Option<usize>,
+    enable_icon: bool,
+    winwidth: Option<usize>,
+) -> Result<()> {
+    use std::path::Path;
 
-    ranked.par_sort_unstable_by(|(_, v1, _), (_, v2, _)| v2.partial_cmp(&v1).unwrap());
+    fn prepend_icon(line: &str) -> String {
+        format!("{} {}", icon_for(line), line)
+    }
+
+    fn icon_for(line: &str) -> &str {
+        let path = Path::new(line);
+        icon_for_path(&path)
+    }
+
+    // Why lazy static? Statics are good only if you would change them during runtime,
+    // but this one looks like a constant lookup table to me.
+    //
+    //XXX TODO: this const tables could get much faster, but there's a lot of work to do:
+    //XXX
+    //XXX 1. This array of `(&str, &str)` tuples should be broken into two arrays:
+    //XXX   [&str] and [Enum], with
+    //XXX   `impl Enum { #[inline] fn into_icon(self) -> &'static str { match self { Variant -> "icon", } } } }`
+    //XXX
+    //XXX 2. The [&str] array should be sorted in alphabetical order, and [Enum] array should
+    //XXX   match sorted filenames or extensions with proper icon at the given index.
+    //XXX
+    //XXX 3. Now, when those filenames and extensions are sorted in alphabetical order,
+    //XXX  we can write this:
+    //XXX
+    //XXX  ```
+    //XXX  let look_for_extension = |slice, file_or_ext| {
+    //XXX      if let Some((idx, _)) = slice.iter().enumerate().find(|&(idx, string)| string == file_or_ext) {
+    //XXX          ENUM_MAP[idx].into_icon()
+    //XXX      } else {
+    //XXX          DEFAULT_ICON
+    //XXX      }
+    //XXX  };
+    //XXX
+    //XXX  let first_letter = match path_part.chars().next() { Some(c) => c, None => return DEFAULT_ICON };
+    //XXX  if first_letter.is_ascii() {
+    //XXX    match first_letter {
+    //XXX      // This match could just return `0..N`, `N..M` (or whatever range it'll generate)
+    //XXX      // right into `look_for_extension(EXTENSION_MAP[matched_range], path_part)`,
+    //XXX      // but verbose example is easier to understand.
+    //XXX      '.' => look_for_extension(EXTENSION_MAP[0..N], path_part),
+    //XXX      'a' => look_for_extension(EXTENSION_MAP[N..M], path_part),
+    //XXX      ...
+    //XXX      _ => DEFAULT_ICON,
+    //XXX    })
+    //XXX  } else {
+    //XXX    // We have no non-ascii files or extensions for now
+    //XXX    DEFAULT_ICON
+    //XXX  }
+    //XXX  ```
+    //XXX
+    //XXX This really looks like a job for some simple script (python, bash),
+    //XXX that will generate a .rs file with the proper Rust code.
+    //XXX Even though this could be easily done manually once, updating such tables
+    //XXX and matches is very prone to errors and it's not something
+    //XXX you want to do, if you can just add one line to script instead.
+    //XXX But if don't plan to update those things at all, that could be done without script once.
+    const EXACTMATCH_MAP: [(&'static str, &'static str); 33] = [
```

Comment:
I have made #354 to split out `src/icon.rs` as another crate. Can you pick out the icon related code into another PR after #354 is merged? That would make this PR cleaner. The icon optimization looks great to me.